README.md
# HelloMinikube

## What is HelloMinikube?
HelloMinikube implements a simple Python application running inside a container
and then deploy in Minikube, which is a local Kubernetes cluster.
This is running **exclusively** on MacOS (Linux supports coming soon!)

## News
* 2019-03-25 - v0.0.2 released! [[download](https://github.com/obutterbach/helloworld_minikube/releases/tag/v0.0.2)]
* 2019-03-25 - v0.0.1 released! [[download](https://github.com/obutterbach/helloworld_minikube/releases/tag/v0.0.1)]

## Quick Start

### Check if all requirements are installed on your machine:

`./preinstall`

This would check and/or install the following requirements:
* Docker for Mac Desktop (require to have a login on Docker Hub)
* Hyperkit
* Minikube


### Build and deploy your app:

`./deploy`

This would implement the following:
* Build helloworld Docker image
* Create a deployment with 2 containers
* Create a service to load balance requests

### How to verify that's running

Start Minikube tunnel (require higher privileges, start in separate terminal):

`minikube tunnel`

This would then allow you to connect to the service, do the following to collect
your cluster information:

`kubectl get services hello-node`

Finally, you should have something as follow display (Cluster IP may differ):

```
20:58 $ kubectl get services hello-node
NAME         TYPE           CLUSTER-IP      EXTERNAL-IP     PORT(S)          AGE
hello-node   LoadBalancer   10.107.43.210   10.107.43.210   8080:31492/TCP   10h
```

To display *Hello World !* message, please curl CLUSTER-IP on port 8080, i.e.:

```
20:59 $ curl 10.107.43.210:8080
Hello World !
```

## Installation
You do have the following options, `preinstall` script or you can follow the official
documentation from Kubernetes website:
* Run `./preinstall.sh` that would guide you through the Minikube installation
* [**Installation**](https://kubernetes.io/docs/tasks/tools/install-minikube/).
Follow the official guideline for installing Minikube


CHANGELOG.md
# HelloMinikube Release Notes

## Version 0.0.2

* Fix deploy and service methods
* Improved README
* Fix server.py by sending proper HTTP status code

## Version 0.0.1

* Add deploy / preinstall Docker folder and dependencies to run a simple
helloworld application


preinstall.sh
#!/usr/bin/env bash
# Preinstall HelloMinikube.
#title          :preinstall.sh
#description    :This script will install Minikube is installed on Mac / Linux
#author         :Oli
#date           :24/03/2019
#version        :0.1
#usage          :bash <(curl -s https://raw.githubusercontent.com/obutterbach/helloworld_minikube/blob/master/preinstall.sh)
#bash_version   :3.2.57(1)-release
#===================================================================================

set -o errexit
set -o pipefail
set -o nounset

RESET_COLOR="\033[0m"
RED_COLOR="\033[0;31m"
GREEN_COLOR="\033[0;32m"
BLUE_COLOR="\033[0;34m"

function reset_color() {
  echo -e "${RESET_COLOR}\c"
}

function red_color() {
  echo -e "${RED_COLOR}\c"
}

function green_color() {
  echo -e "${GREEN_COLOR}\c"
}

function blue_color() {
  echo -e "${BLUE_COLOR}\c"
}

function separator() {
  green_color
  echo "#=============================STEP FINISHED=============================#"
  reset_color
}

function hello() {
  blue_color
  echo "This script will guide you through installing all required dependencies"
  echo "for installing Minikube"
  echo "This script requires elevated privileges"
  green_color
  read -p "Do you want to proceed with installation? (y/N) " -n 1 answer
  echo
  if [ ${answer} != "y" ]; then
      exit 1
  fi
}

function install_docker() {
  blue_color
  echo "Trying to detect installed Docker..."
  if ! [ $(command -v docker 2>&1) ]; then
    blue_color
    echo "You don't have Docker installed"
    echo "This is required to proceed with installation"
    echo "Please follow the instruction at the following location"
    echo "https://docs.docker.com/docker-for-mac/install/"
  else
    blue_color
    echo "Seems like you have installed Docker, so skipping..."
  fi
  reset_color
  separator
  sleep 1
}

function install_hyperkit_driver() {
  blue_color
  echo "Trying to detect installed Hyperkit driver..."
  if ! [ $(command -v hyperkit 2>&1) ]; then
    blue_color
    echo "You don't have Hyperkit driver installed"
    echo "This is required to proceed with installation"
    green_color
    read -p "Do you agree to install Hyperkit driver? (y/N) " -n 1 answer
    echo
    if [ ${answer} != "y" ]; then
        exit 1
    fi
    blue_color
    echo "Installing Hyperkit driver..."
    echo "Please, wait until Hyperkit driver will be installed, before continue"
    curl -LO https://storage.googleapis.com/minikube/releases/latest/docker-machine-driver-hyperkit \
    && sudo install -o root -g wheel -m 4755 docker-machine-driver-hyperkit /usr/local/bin/hyperkit \
    && rm docker-machine-driver-hyperkit
  else
    blue_color
    echo "Seems like you have installed Hyperkit driver, so skipping..."
  fi
  reset_color
  separator
  sleep 1
}

function install_minikube() {
  blue_color
  echo "Trying to detect installed Minikube..."
  if ! [ $(command -v minikube 2>&1) ]; then
    blue_color
    echo "You don't have Minikube installed"
    echo "This is required to proceed with installation"
    green_color
    read -p "Do you agree to install Minikube? (y/N) " -n 1 answer
    echo
    if [ ${answer} != "y" ]; then
        exit 1
    fi
    blue_color
    echo "Installing Minikube..."
    echo "Please, wait until Minikube will be installed, before continue"
    curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube-darwin-amd64 \
    && sudo install minikube-darwin-amd64 /usr/local/bin/minikube \
    && rm minikube-darwin-amd64
  else
    blue_color
    echo "Seems like you have installed Minikube, so skipping..."
  fi
  reset_color
  separator
  sleep 1
}

function start_minikube() {
  blue_color
  echo "Make sure Minikube is running..."
  minikube start --vm-driver hyperkit
  reset_color
  separator
  sleep 1
}

hello
install_docker
install_hyperkit_driver
install_minikube
start_minikube


Docker/server.py
#!/usr/bin/env python3
from http.server import BaseHTTPRequestHandler,HTTPServer

PORT_NUMBER = 8080

# This class will handle any incoming request from
# a browser
class myHandler(BaseHTTPRequestHandler):
    # Handler for the GET requests
    def do_GET(self):
        print('Get request received')
        self.send_response(200)
        self.send_header('Content-type','text/html')
        self.end_headers()
        # Send the html message
        self.wfile.write("Hello World !".encode())
        return

try:
    # Create a web server and define the handler to manage the
    # incoming request
    server = HTTPServer(('', PORT_NUMBER), myHandler)
    print ('Started httpserver on port ' , PORT_NUMBER)

    # Wait forever for incoming http requests
    server.serve_forever()

except KeyboardInterrupt:
    print ('^C received, shutting down the web server')
    server.socket.close()


deploy.sh
#!/usr/bin/env bash
# Deploy HelloMinikube.
#title          :deploy.sh
#description    :This script will install Minikube is installed on Mac / Linux
#author         :Oli
#date           :25/03/2019
#version        :0.1
#usage          :bash <(curl -s https://raw.githubusercontent.com/obutterbach/helloworld_minikube/blob/master/deploy.sh)
#bash_version   :3.2.57(1)-release
#===================================================================================

set -o errexit
set -o pipefail
set -o nounset

# Make sure to run docker with Minikube
eval $(minikube docker-env)

RESET_COLOR="\033[0m"
RED_COLOR="\033[0;31m"
GREEN_COLOR="\033[0;32m"
BLUE_COLOR="\033[0;34m"

function reset_color() {
  echo -e "${RESET_COLOR}\c"
}

function red_color() {
  echo -e "${RED_COLOR}\c"
}

function green_color() {
  echo -e "${GREEN_COLOR}\c"
}

function blue_color() {
  echo -e "${BLUE_COLOR}\c"
}

function separator() {
  green_color
  echo "#=============================STEP FINISHED=============================#"
  reset_color
}

function build_container() {
  blue_color
  echo "Building Docker container..."
  green_color
  docker build -t hello-world-python Docker/
  reset_color
  separator
  sleep 1
}

function create_deployment() {
  blue_color
  echo "Create deployment..."
  green_color
  kubectl get deployments hello-node || kubectl run hello-node --replicas=2 --image=hello-world-python:latest --port=8080 --image-pull-policy=Never
  reset_color
  separator
  sleep 1
}

function create_service() {
  blue_color
  echo "Create deployment..."
  green_color
  kubectl get services hello-node || kubectl expose deployment hello-node --type=LoadBalancer --port=8080
  reset_color
  separator
  sleep 1
}

build_container
create_deployment
create_service
